File: C/Binary Search Tree/BST.c
#include "BST.h"

// Constructor
BST *create_tree()
{
    BST *tree = (BST *)malloc(sizeof(BST));
    tree->_root = NULL;
    tree->_size = 0;
    return tree;
}

// Destructor
void destroy_tree(BST *tree)
{
    destroy_node(tree->_root);
    free(tree);
}

// Methods
BST_Node *insert(BST *tree, BST_Node *root, int data)
{
    if (data == root->_data)
    {
        return root;
    }

    if (root == NULL)
    {
        tree->_size++;
        return create_node(data);
    }
    else if (data > root->_data)
    {
        root->_right = insert(tree, root->_right, data);
    }
    else if (data < root->_data)
    {
        root->_left = insert(tree, root->_left, data);
    }
}

BST_Node *find_ios(BST *tree, BST_Node *root)
{
    while (root->_left != NULL)
    {
        root = root->_left;
    }

    return root;
}

BST_Node *remove(BST *tree, BST_Node *root, int data)
{
    if (root == NULL)
    {
        return root;
    }

    if (root->_data == data)
    {
        // Case 1: No children
        if (root->_left == NULL && root->_right == NULL)
        {
            destroy_node(root);
            return NULL;
        }

        // Case 2: One child
        if (root->_left == NULL)
        {
            BST_Node *temp = root->_right;
            root->_right = remove(tree, root->_right, temp->_data);
            return temp;
        }
        if (root->_right == NULL)
        {
            BST_Node *temp = root->_left;
            root->_left = remove(tree, root->_left, temp->_data);
            return temp;
        }

        // Case 3: Two children
        BST_Node *ios = find_ios(tree, root->_right);
        root->_data = ios->_data;
        root->_right = remove(tree, root->_right, ios->_data);
    }
    else if (data > root->_data)
    {
        root->_right = remove(tree, root->_right, data);
    }
    else if (data < root->_data)
    {
        root->_left = remove(tree, root->_left, data);
    }

    return root;
}

bool search(BST *tree, BST_Node *root, int data)
{
    if (root == NULL)
    {
        return false;
    }

    if (root->_data == data)
    {
        return true;
    }
    else if (data > root->_data)
    {
        return search(tree, root->_right, data);
    }
    else if (data < root->_data)
    {
        return search(tree, root->_left, data);
    }
}

void preorder(BST *tree, BST_Node *root)
{
    if (root == NULL)
    {
        return;
    }

    printf(root->_data + " ");
    preorder(tree, root->_left);
    preorder(tree, root->_right);
}

void inorder(BST *tree, BST_Node *root)
{
    if (root == NULL)
    {
        return;
    }

    inorder(tree, root->_left);
    printf(root->_data + " ");
    inorder(tree, root->_right);
}

void postorder(BST *tree, BST_Node *root)
{
    if (root == NULL)
    {
        return;
    }

    postorder(tree, root->_left);
    postorder(tree, root->_right);
    printf(root->_data + " ");
}

int height(BST *tree, BST_Node *root)
{
    if (root == NULL)
    {
        return 0;
    }

    int left_height = height(tree, root->_left);
    int right_height = height(tree, root->_right);

    return (left_height > right_height) ? left_height + 1 : right_height + 1;
}

File: Java/binary_search_tree/BST.java
package binary_search_tree;

public class BST {
    private BSTNode root;

    // Constructor
    public BST() {
        this.root = null;
    }

    // Private Methods
    private BSTNode insert(BSTNode current, int data) {
        if (current == null) {
            return new BSTNode(data);
        }

        if (data < current.getData()) {
            current.setNodeLeft(this.insert(current.getNodeLeft(), data));
        } else {
            current.setNodeRight(this.insert(current.getNodeRight(), data));
        }

        return current;
    }

    private BSTNode remove(BSTNode current, int data) {
        if (current == null) {
            return null;
        }

        if (current.getData() == data) {
            // 2 children
            if (current.getNodeRight() != null && current.getNodeLeft() != null) {
                BSTNode ios = this.findIOS(current.getNodeRight());
                current.setData(ios.getData());
                current.setNodeRight(this.remove(current.getNodeRight(), ios.getData()));
            }

            // 1 child (right)
            else if (current.getNodeRight() != null) {
                current = current.getNodeRight();
            }
            // 1 child (left)
            else if (current.getNodeLeft() != null) {
                current = current.getNodeLeft();
            }

            // 0 children
            else {
                current = null;
            }
        } else if (data < current.getData()) {
            current.setNodeLeft(this.remove(current.getNodeLeft(), data));
        } else {
            current.setNodeRight(this.remove(current.getNodeRight(), data));
        }

        return current;
    }

    private int height(BSTNode root) {
        if (root == null) {
            return 0;
        }

        int leftHeight = this.height(root.getNodeLeft());
        int rightHeight = this.height(root.getNodeRight());

        // Returns the greater of the two heights
        if (leftHeight > rightHeight) {
            return leftHeight + 1;
        } else {
            return rightHeight + 1;
        }
    }

    private String preorder(BSTNode root) {
        String outString = "";

        if (root != null) {
            outString += root.getData() + " ";
            outString += this.preorder(root.getNodeLeft());
            outString += this.preorder(root.getNodeRight());
        }

        return outString;
    }

    private String inorder(BSTNode root) {
        String outString = "";

        if (root != null) {
            outString += this.inorder(root.getNodeLeft());
            outString += root.getData() + " ";
            outString += this.inorder(root.getNodeRight());
        }

        return outString;
    }

    private String postorder(BSTNode root) {
        String outString = "";

        if (root != null) {
            outString += this.postorder(root.getNodeLeft());
            outString += this.postorder(root.getNodeRight());
            outString += root.getData() + " ";
        }

        return outString;
    }

    private BSTNode findIOS(BSTNode root) {
        while (root.getNodeLeft() != null) {
            root = root.getNodeLeft();
        }

        return root;
    }

    // Public Methods
    public void insert(int data) {
        if (!this.search(data)) {
            this.root = this.insert(this.root, data);
        }
    }

    public int height() {
        if (this.root == null) {
            return -1;
        }

        return this.height(this.root);
    }

    public void remove(int data) {
        if (this.search(data)) {
            this.root = this.remove(this.root, data);
        }
    }

    public void preorder() {
        System.out.println(this.preorder(this.root));
    }

    public void inorder() {
        System.out.println(this.inorder(this.root));
    }

    public void postorder() {
        System.out.println(this.postorder(this.root));
    }

    public boolean search(int data) {
        BSTNode current = this.root;
        while (current != null) {
            if (current.getData() == data) {
                return true;
            } else if (data < current.getData()) {
                current = current.getNodeLeft();
            } else {
                current = current.getNodeRight();
            }
        }
        return false;
    }
}